sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -0,0 +1,215 @@
+import time
+import queue
+import logging
+from threading import RLock, Condition, Semaphore
+from concurrent.futures import ThreadPoolExecutor
+from typing import Optional, Callable, TYPE_CHECKING
+
+from .._producer import EventHubProducer
+from .._common import EventDataBatch
+from ..exceptions import OperationTimeoutError
+
+if TYPE_CHECKING:
+    from .._producer_client import SendEventTypes
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BufferedProducer:
+    # pylint: disable=too-many-instance-attributes
+    def __init__(
+            self,
+            producer: EventHubProducer,
+            partition_id: str,
+            on_success: Callable[["SendEventTypes", Optional[str]], None],
+            on_error: Callable[["SendEventTypes", Optional[str], Exception], None],
+            max_message_size_on_link: int,
+            executor: ThreadPoolExecutor,
+            *,
+            max_wait_time: float = 1,
+            max_concurrent_sends: int = 1,
+            max_buffer_length: int = 10
+    ):
+        self._buffered_queue: queue.Queue = queue.Queue()
+        self._cur_buffered_len = 0
+        self._executor: ThreadPoolExecutor = executor
+        self._producer: EventHubProducer = producer
+        self._lock = RLock()
+        self._not_empty = Condition(self._lock)
+        self._not_full = Condition(self._lock)
+        self._max_buffer_len = max_buffer_length
+        self._max_concurrent_sends = max_concurrent_sends
+        self._max_concurrent_sends_semaphore = Semaphore(self._max_concurrent_sends)
+        self._max_wait_time = max_wait_time
+        self._on_success = self.enhanced_callback_decorator(on_success)
+        self._on_error = self.enhanced_callback_decorator(on_error)
+        self._last_send_time = None
+        self._running = False
+        self._cur_batch: Optional[EventDataBatch] = None
+        self._max_message_size_on_link = max_message_size_on_link
+        self._check_max_wait_time_future = None
+        self.partition_id = partition_id
+
+    def start(self):
+        with self._lock:
+            self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+            self._buffered_queue.put(self._cur_batch)
+            self._running = True
+            if self._max_wait_time:
+                self._last_send_time = time.time()
+                self._check_max_wait_time_future = self._executor.submit(self.check_max_wait_time_worker)
+
+    def stop(self, flush=True, timeout=None, raise_error=False):
+        self._running = False
+        if self._check_max_wait_time_future:
+            try:
+                self._check_max_wait_time_future.result(timeout)
+            except Exception as exc:  # pylint: disable=broad-except
+                _LOGGER.warning(
+                    "Partition %r stopped with error %r",
+                    self.partition_id,
+                    exc
+                )
+        if flush:
+            self.flush(timeout=timeout, raise_error=raise_error)
```

Comment:
I've re-ordered the two step and I realize that it's actually the right thing to flush first then wait for `check_max_wait_time_future` to finish.

otherwise the close turns out to be a double flush operation which is unnecessary.